### DISCUSSION TITLE
Not Recognizing Implemented Provisions?

### DISCUSSION THREAD

**Discussion by MGatner** [composer-unused/composer-unused#234] [Q&A]
Hi Composer Unused- I have a project using `knplabs/github-api` which requires a `psr/http-factory-implementation`. I am fulfilling that with `guzzlehttp/guzzle` via `http-interop/http-factory-guzzle`, however both of those are being detected as "unused" ([for example](https://github.com/lonnieezell/ci-website/runs/4093396449?check_suite_focus=true)). Is that because of the intermediary step?

The code is [definitely using the KNPLabs Client](https://github.com/lonnieezell/ci-website/blob/develop/app/Libraries/GitHub.php#L11) and the Client is [definitely using the HTTP layer](https://github.com/KnpLabs/php-github-api/blob/de2f27846139e7b4b92dcad9ff77660d0dc5dfad/lib/Github/Client.php#L14).

**Comment by icanhazstring** [comment]
Hi. This could have multiple issues. Let me check on that in detail and I get back to you.

In the meantime you could check the `dev-main` version of this problems persists. I changed quite some things for the next version, maybe it's already fixed.

**Reply by MGatner** [reply]
Thanks! I will try that.

**Reply by icanhazstring** [reply]
I think the problem might still persist.
Those virtual packages `-implementation` are tricky ones here.

I do scan for the following:
- direct required and used dependencies
- direct required and required by 3rd party
- direct required and suggested by 3rd party

But the missing one, which is the problem here:
- direct required and used by 3rd party

So what might need to be done is the following:
Lets say you have 3 packages. A, B and C

Root requires: A, B, C
Root uses: A and B (directly)

That leaves C without usage. BUT!
If C is actually used by any of A or B while those are used by Root. Package C should be marked used as well.

**Reply by icanhazstring** [reply]
I just created an Issue for that topic. Thanks for bringing that up :)
https://github.com/composer-unused/composer-unused/issues/235
